[Pack] Pack Pattern Root IPIN

amin1377 opened the pull request:
WIP

AmirhosseinPoolad commented on vpr/src/pack/prepack.cpp:
```diff
@@ -1601,9 +1655,11 @@ static void get_all_connected_primitive_pins(const t_pb_graph_pin* cluster_input
         for (int ipin = 0; ipin < output_edge->num_output_pins; ipin++) {
             if (output_edge->output_pins[ipin]->is_primitive_pin()) {
-                connected_primitive_pins.push_back(output_edge->output_pins[ipin]);
+                if (pattern_blocks.find(output_edge->output_pins[ipin]->parent_node->pb_type) != pattern_blocks.end()) {
```
It's also not obvious at a glance what it would mean for pattern_blocks to contain the pb_type of output_pins[ipin]. You could add a comment here to help with that

amin1377 replied:
I actually considered using contains instead of find, but since I want this code to be compatible with OpenFPGA — which isn’t compiled with C++20 — I decided against it. Also, I think std::unordered_set::contains is just more expressive than find and in terms of performance, it is not different from std::find.

amin1377 replied:
added the followign comment:

```
/** Add the output pin to the vector only if it belongs to a pb_type registered in the pattern_blocks set */
```

AmirhosseinPoolad replied:
Makes sense. I guess the only thing remaining is that you used find here and count in the other place to do this thing. I think it's better to just one method for consistency. Don't have a particular preference, use whichever method you like more.

amin1377 replied:
Good point. I changed the counts to find.

AmirhosseinPoolad commented on vpr/src/pack/prepack.cpp:
```diff
@@ -1172,6 +1182,47 @@ static AtomBlockId get_driving_block(const AtomBlockId block_id,
+static std::unordered_set<t_pb_type*> get_pattern_blocks(const t_pack_patterns* pack_pattern) {
+    std::unordered_set<t_pb_type*> pattern_blocks;
+
+    auto connections = pack_pattern->root_block->connections;
```
Change auto to actual type name

amin1377 replied:
You’re really not a fan of auto :)

AmirhosseinPoolad commented on vpr/src/pack/prepack.cpp:
```diff
@@ -1172,6 +1182,47 @@ static AtomBlockId get_driving_block(const AtomBlockId block_id,
+    pack_pattern_blocks.push(connections->from_block);
+    /** Start from the root block of the pack pattern and add the connected block to the queue */
+    while (!pack_pattern_blocks.empty()) {
+        auto current_pattern_block = pack_pattern_blocks.front();
```
Same as above

AmirhosseinPoolad commented on vpr/src/pack/prepack.cpp:
```diff
@@ -1172,6 +1182,47 @@ static AtomBlockId get_driving_block(const AtomBlockId block_id,
+    while (!pack_pattern_blocks.empty()) {
+        auto current_pattern_block = pack_pattern_blocks.front();
+        pack_pattern_blocks.pop();
+        auto current_connenction = current_pattern_block->connections;
```
Auto

AmirhosseinPoolad commented on vpr/src/pack/prepack.cpp:
```diff
@@ -1172,6 +1182,47 @@ static AtomBlockId get_driving_block(const AtomBlockId block_id,
+         * add the connected block to the queue
+         */
+        while (current_connenction != nullptr) {
+            if (visited_from_pins.count(current_connenction->from_pin)) {
+                if (visited_to_pins.count(current_connenction->to_pin)) {
```
These two can use std::unordered_set::contains instead of count.

amin1377 replied:
Same as I mentioned above, but let me know if that’s not convincing, and I’m happy to switch to contains instead of find.

AmirhosseinPoolad commented on vpr/src/pack/prepack.cpp:
```diff
@@ -117,6 +119,14 @@ static AtomBlockId get_driving_block(const AtomBlockId block_id,
+ * @param pack_pattern Pack pattern to get pb_types from
+ * @return std::unordered_set<t_pb_type*> Set of pb_types in the pack pattern
+ */
+static std::unordered_set<t_pb_type*> get_pattern_blocks(const t_pack_patterns* pack_pattern);
```
Nitpick: why not use a reference to pass pack_pattern? This part of the code unfortunately uses a ton of C-style arrays so using the reference syntax makes it super clear that it's not an array.

amin1377 replied:
You’re right, it makes more sense to pass it as a const reference. I was trying to stay consistent with other parts of the code, but I agree that it’s better to follow modern C++ style for new code. We can always refactor the older parts later for consistency.

AmirhosseinPoolad reviewed (changes requested):
Thanks for the changes Amin! I'll let Vaughn comment on the actual functionality changes, but meanwhile I had a couple of style related comments and nitpicks.

amin1377 commented on vpr/src/pack/prepack.cpp:
```diff
@@ -1590,7 +1642,9 @@ static void update_chain_root_pins(t_pack_patterns* chain_pattern,
  *  the Cin pin of all the adder primitives connected to this pin. Which is for typical architectures
  *  will be only one pin connected to the very first adder in the cluster.
  */
-static void get_all_connected_primitive_pins(const t_pb_graph_pin* cluster_input_pin, std::vector<t_pb_graph_pin*>& connected_primitive_pins) {
+static void get_all_connected_primitive_pins(const t_pb_graph_pin* cluster_input_pin,
+                                             const std::unordered_set<t_pb_type*>& pattern_blocks,
+                                             std::vector<t_pb_graph_pin*>& connected_primitive_pins) {
```
Added the following comment:

```
/**
 * @brief Get all primitive pins connected to the given cluster input pin
 * 
 * @param cluster_input_pin Cluster input pin to get connected primitive pins from
 * @param pattern_blocks Set of pb_types in the pack pattern. Pins on the blocks in this set will 
 * be added to the connected_primitive_pins vector
 * @param connected_primitive_pins Vector to store connected primitive pins
 */
```

amin1377 commented:
@AmirhosseinPoolad Thanks for reviewing the code! I’ve addressed your comments. Please resolve the conversations if you think they’ve been properly handled. Let me know if you have any further suggestions.

soheilshahrouz commented on vpr/src/pack/prepack.cpp:
```diff
@@ -1172,6 +1190,47 @@ static AtomBlockId get_driving_block(const AtomBlockId block_id,
+        t_pack_pattern_connections* current_connenction = current_pattern_block->connections;
+        /** Iterate through all the connections of the current pattern block to
+         * add the connected block to the queue
+         */
```
Comment blocks starting with `/**` are parsed as doxygen comments by some IDEs.

soheilshahrouz commented on vpr/src/pack/prepack.cpp:
```diff
@@ -1172,6 +1190,47 @@ static AtomBlockId get_driving_block(const AtomBlockId block_id,
+        t_pack_pattern_block* current_pattern_block = pack_pattern_blocks.front();
+        pack_pattern_blocks.pop();
+        t_pack_pattern_connections* current_connenction = current_pattern_block->connections;
+        /** Iterate through all the connections of the current pattern block to
```
Adding some empty lines can make the code more readable, especially when added between different logical steps that are taken in the code.

AlexandreSinger reviewed (approved):
LGTM
